docs/design/copy-info.md

```diff
@@ -0,0 +1,157 @@
+no obligation to do so, as clients are expected to use the `get_copy_records()`
+API to get copy info instead. A backend which does not support explicit copy
+info may choose to silently ignore it or return an error.
+
+This API will enable the creation of new `jj` commands for recording copies:
+
+```shell
+jj cp $SRC $DEST [OPTIONS]
+jj mv $SRC $DEST [OPTIONS]
+```
+
+These commands will perform the filesystem operations associated with
+copies/moves, as well as editing the working copy commit to include records of
+these copies.
+
+Flags for both commands will include:
+
+```
+-f
+    force overwrite the destination path
+--after
+    record the copy retroactively, without modifying the filesystem
+--from REV
+    specify a commit id for the copy source that isn't the parent commit
+```
+
+For backends which do not support explicit copies, it will be an error to use
+`--after`, since this has no effect on anything and the user should know that.
+
+### Rebase Changes
+
+A well known and thorny problem in Mercurial occurs in the following scenario:
+
+1.  Create a new file A
+1.  Create new commits on top that make changes to file A
+1.  Whoops, I should rename file A to B. Do so, amend the first commit.
+1.  Because the first commit created file A, there is no rename to record; it's changing to a commit that instead creates file B.
+1.  All child commits get sad on evolve
+
+In jj, we have an opportunity to fix this because all rebasing occurs atomically
+and transactionally within memory. The exact implementation of this is yet to be
+determined, but conceptually the following should produce desirable results:
+
+1.  Rebase a commit A from parents [B] to parents [C]
+1.  Get copy records from [D]->[B] and [D]->[C], where [D] are the common ancestors of [B] and [C]
+1.  DescendantRebaser maintains an in-memory map of commits to extra copy info, which it may inject into (2). When squashing a rename of a newly created file into the commit that creates that file, DescendentRebase will return this rename for all rebases of descendants of the newly modified commit. The rename lives ephemerally in memory and has no persistence after the rebase completes.
+1.  A to-be-determined algorithm diffs the copy records between [D]->[B] and [D]->[C] in order to make changes to the rebased commit. This results in edits to renamed files being propagated to those renamed files, and avoiding conflicts on the deletion of their sources. A copy/move may also be undone in this way; abandoning a commit which renames A->B should move all descendant edits of B back into A.
+
+In general, we want to support the following use cases:
+
+-   A rebase of an edited file A across a rename of A->B should transparently move the edits to B.
+-   A rebase of an edited file A across a copy from A->B should _optionally_ copy the edits to B. A configuration option should be defined to enable/disable this behavior.
+-   TODO: Others?
```

Comment:
I think there are many strange and interesting merge cases to consider. For example:

```
C files: R->S=a
|
|  B files: R=b, S=b
|/
A files: R=a
```

Since commit `B` changed the contents of file `R` from `a` to `b` and commit `C` renamed file `R` to `S` (without changing its contents), the result of those two operations would be file `S` with contents `b`. Does that conflict with file `S` (with the same contents) that was added in commit `B`?

Reply:
I think the new conflicts section covers this.